icu4c/source/io/ufile.cpp

```diff
@@ -142,39 +143,56 @@ u_fopen(const char    *filename,
     return result; /* not a file leak */
 }
 
+enum { kFilenameBufCapacity = 256 };
+
 U_CAPI UFILE* U_EXPORT2
 u_fopen_u(const UChar   *filename,
         const char    *perm,
         const char    *locale,
         const char    *codepage)
 {
-    UFILE     *result;
-    char buffer[256];
-
-    u_austrcpy(buffer, filename);
+    UFILE *result;
+    char buffer[kFilenameBufCapacity];
+    char *filenameBuffer = buffer;
+
+    icu::UnicodeString filenameString(true, filename, -1); // readonly aliasing, does not allocate memory
+    // extract with conversion to platform default codepage, return full length (not including 0 termination)
+    int32_t filenameLength = filenameString.extract(0, filenameString.length(), filenameBuffer, kFilenameBufCapacity);
+    if (filenameLength >= kFilenameBufCapacity) { // could not fit (with zero termination) in buffer
+        filenameBuffer = static_cast<char *>(uprv_malloc(++filenameLength)); // add one for zero termination
+        if (!filenameBuffer) {
+            return nullptr;
+        }
+        filenameString.extract(0, filenameString.length(), filenameBuffer, filenameLength);
+    }
 
-    result = u_fopen(buffer, perm, locale, codepage);
+    result = u_fopen(filenameBuffer, perm, locale, codepage);
 #if U_PLATFORM_USES_ONLY_WIN32_API
     /* Try Windows API _wfopen if the above fails. */
     if (!result) {
         // TODO: test this code path, including wperm.
         wchar_t wperm[40] = {};
         size_t  retVal;
         mbstowcs_s(&retVal, wperm, UPRV_LENGTHOF(wperm), perm, _TRUNCATE);
-        FILE *systemFile = _wfopen((const wchar_t *)filename, wperm);
+        FILE *systemFile = _wfopen((const wchar_t *)filename, wperm); // may return NULL for filename > 255 chars
```

Comment:
nit: I know it's existing code... but this function now mixes C style casts with C++ style casts. :) 
Would you mind changing this to a `reinterpret_cast<>` ?